React 18 RC - Upgrade Guide (#4384)
reactjs/reactjs.org: content/blog/2022-02-25-react-v18-rc.md

function handleClick() {
  setCount(c => c + 1);
  setFlag(f => !f);
  // React will only re-render once at the end (that's batching!)
}

setTimeout(() => {
  setCount(c => c + 1);
  setFlag(f => !f);
  // React will only re-render once at the end (that's batching!)
}, 1000);
```

This is a breaking change, but we expect this to result in less work rendering, and therefore better performance in your applications. To opt-out of automatic batching, you can use `ReactDOM.flushSync`:

```js
function handleClick() {
  flushSync(() => {
    setCounter(c => c + 1);
  });
  // React has updated the DOM by now
  flushSync(() => {
    setFlag(f => !f);
  });
  // React has updated the DOM by now
}
```

For more information, see the [Automatic batching deep dive](https://github.com/reactwg/react-18/discussions/21).

## New APIs for libraries

In the React 18 Working Group we worked with library maintainers to create new APIs needed to support concurrent rendering for use cases specific to their use case in areas like styles, external stores, and accessibility. To support React 18, some libraries may need to switch to one of the following APIs:

> * `useID` is a new hook for generating unique IDs on both the client and server, while avoiding hydration mismatches. This solves an issue that already exists in React 17 and below, but it's even more important in React 18 because of how our streaming server renderer delivers HTML out-of-order. For more information see the [useId post in the working group](https://github.com/reactwg/react-18/discussions/111).

Reviewer: ```suggestion
* `useId` is a new hook for generating unique IDs on both the client and server, while avoiding hydration mismatches. This solves an issue that already exists in React 17 and below, but it's even more important in React 18 because of how our streaming server renderer delivers HTML out-of-order. For more information see the [useId post in the working group](https://github.com/reactwg/react-18/discussions/111).
```
[Or shall we do a round 😆 ?](https://discord.com/channels/845008699655323698/845008699655323701/904859805590634497)

---
title: "React v18.0 Release Candidate: Upgrade Guide"
author: [rickhanlonii]
---

> Today we’re publishing the second release candidate for React 18. With this release, we’re encouraging users to install and test React 18 to [raise any issues](https://github.com/facebook/react/issues) for any problems you encounter in the upgrade before we publish the full stable release.

Reviewer: nit: since this is about "raising issues" let's directly link the "new issue" link:
```suggestion
Today we’re publishing the second release candidate for React 18. With this release, we’re encouraging users to install and test React 18 to [raise any issues](https://github.com/facebook/react/issues/new/choose) for any problems you encounter in the upgrade before we publish the full stable release.
```

Ideally we could link the "React 18" issue but that would hardcode the templates, labels etc. So once you change the template or label, the link in this repo would be outdated.

As we shared at [React Conf](https://reactjs.org/blog/2021/12/17/react-conf-2021-recap.html), React 18 will include our new concurrent renderer and a gradual migration strategy for concurrent features. React 18 also includes a small number of breaking changes and features outside of concurrent rendering. In this post we will describe the changes to expect when upgrading to React 18.

*Note for React Native users: React 18 will ship in React Native with the New React Native Architecture. For more information, see the [React Conf keynote here](https://www.youtube.com/watch?v=FZ0cG47msEk&t=1530s).*

## Installing

To install the latest React 18 RC, use the `@rc` tag:

```bash
npm install react@rc react-dom@rc
```

Or if you’re using yarn:

```bash
yarn add react@rc react-dom@rc
```

## New client root APIs

When you first install React 18, you will see a warning in the console:

> ![Screenshot of React error with text "Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17."](../images/blog/react-v18-rc/reactdom_warning.png)

Reviewer: Alt text 🥳

React 18 introduces a new root API which provides better ergonomics for managing roots. The new root API also enables the new concurrent renderer, which allows you to opt-into concurrent features.

To upgrade `ReactDOM.render`:

```js
// Before
const container = document.getElementById('app');
ReactDOM.render(<App tab="home" />, container);

> unmountComponentAtNode(containter); // ...some time later

Reviewer: We already have a section later for this. It's also not mirrored in the corresponding "After" section.

Author: Nice catch, I was originally going to roll the examples together.

// After
const container = document.getElementById('app');
const root = ReactDOM.hydrateRoot(container, <App tab="home" />);
// Unlike with createRoot, you don't need a separate root.render() call here.
```

For more information, see the [working group discussion here](https://github.com/reactwg/react-18/discussions/5).

## Updates to server rendering APIs

In this release, we’re revamping our server rendering APIs to support Suspense on the server and Streaming SSR.

As part of these changes, the following APIs will continue working, but with limited support for Suspense. Instead of erroring, these APIs will server render the nearest fallback for any subtree that suspends, and retry rendering on the client:

* `renderToString`
* `renderToStaticMarkup`

For full support of Suspense on the server, React 18 introduces a new recommended API. This API will retry suspended Suspense boundaries on the server, streaming the server-rendered content to the client when it is available:

* `renderToPipeableStream`

This new streaming API replaces the old streaming APIs, which are now deprecated and will warn:

* `renderToNodeStream`
* `renderToStaticNodeStream`

> For more information on the changes to these APIs, see the working group post [here](https://github.com/reactwg/react-18/discussions/22), a deep dive on the new Suspense SSR Architecture [here](https://github.com/reactwg/react-18/discussions/37), and [Shaundai Person’s](https://twitter.com/shaundai) talk on [Streaming Server Rendering with Suspense](https://www.youtube.com/watch?v=pj5N-Khihgc) at React Conf 2021.

Reviewer: Avoiding "click-here" links (explainer: https://heyoka.medium.com/dont-use-click-here-f32f445d1021)

```suggestion
For more information on the changes to these APIs, see the working group post on [Upgrading to React 18 on the server](https://github.com/reactwg/react-18/discussions/22), a [deep dive on the new Suspense SSR Architecture](https://github.com/reactwg/react-18/discussions/37), and [Shaundai Person’s](https://twitter.com/shaundai) talk on [Streaming Server Rendering with Suspense](https://www.youtube.com/watch?v=pj5N-Khihgc) at React Conf 2021.
```

  // React will render twice, once for each state update (no batching)
}, 1000);
```

Starting in React 18 with `createRoot`, all updates will be automatically batched, no matter where they originate from. This means that updates inside of timeouts, promises, native event handlers or any other event will batch the same way as updates inside of React events:

```js
// After React 18 updates inside of timeouts, promises,
// native event handlers or any other event are batched.

function handleClick() {
  setCount(c => c + 1);
  setFlag(f => !f);
  // React will only re-render once at the end (that's batching!)
}

setTimeout(() => {
  setCount(c => c + 1);
  setFlag(f => !f);
  // React will only re-render once at the end (that's batching!)
}, 1000);
```

This is a breaking change, but we expect this to result in less work rendering, and therefore better performance in your applications. To opt-out of automatic batching, you can use `ReactDOM.flushSync`:

```js
function handleClick() {
  flushSync(() => {
    setCounter(c => c + 1);
  });
  // React has updated the DOM by now
  flushSync(() => {
    setFlag(f => !f);
  });
  // React has updated the DOM by now
}
```

> For more information, see the deep dive in the [working group here.](https://github.com/reactwg/react-18/discussions/21)

Reviewer: Avoiding "click-here" links
```suggestion
For more information, see the [Automatic batching deep dive](https://github.com/reactwg/react-18/discussions/21).
```

## New APIs for libraries

In the React 18 Working Group we worked with library maintainers to create new APIs needed to support concurrent rendering for use cases specific to their use case in areas like styles, external stores, and accessibility. To support React 18, some libraries may need to switch to one of the following APIs:

> * `useID` is a new hook for generating unique IDs on both the client and server, while avoiding hydration mismatches. This solves an issue that already exists in React 17 and below, but it's even more important in React 18 because of how our streaming server renderer delivers HTML out-of-order. For more information see the post [here](https://github.com/reactwg/react-18/discussions/111).

Reviewer: Avoiding "click-here" links
```suggestion
* `useID` is a new hook for generating unique IDs on both the client and server, while avoiding hydration mismatches. This solves an issue that already exists in React 17 and below, but it's even more important in React 18 because of how our streaming server renderer delivers HTML out-of-order. For more information see the [useId post in the working group](https://github.com/reactwg/react-18/discussions/111).
```

> * `useSyncExternalStore` is a new hook that allows external stores to support concurrent reads by forcing updates to the store to be synchronous. This new API is recommended for any library that integrates with state external to React. For more information, see the overview [here](https://github.com/reactwg/react-18/discussions/70) and API details [here](https://github.com/reactwg/react-18/discussions/86).

Reviewer: Avoiding "click-here" links
```suggestion
* `useSyncExternalStore` is a new hook that allows external stores to support concurrent reads by forcing updates to the store to be synchronous. This new API is recommended for any library that integrates with state external to React. For more information, see the [useSyncExternalStore overview post](https://github.com/reactwg/react-18/discussions/70) and [useSyncExternalStore API details](https://github.com/reactwg/react-18/discussions/86).
```

> * `useInsertionEffect` is a new hook that allows CSS-in-JS libraries to address performance issues of injecting styles in render. This hook will run after the DOM is mutated, but before layout effects read the new layout. This solves an issue that already exists in React 17 and below, but is even more important in React 18 because React yields to the browser during concurrent rendering, giving it a chance to recalculate layout. For more information, see the post [here](https://github.com/reactwg/react-18/discussions/110).

Reviewer: ```suggestion
* `useInsertionEffect` is a new hook that allows CSS-in-JS libraries to address performance issues of injecting styles in render. This hook will run after the DOM is mutated, but before layout effects read the new layout. This solves an issue that already exists in React 17 and below, but is even more important in React 18 because React yields to the browser during concurrent rendering, giving it a chance to recalculate layout. For more information, see the [Library Upgrade Guide for `<style>`](https://github.com/reactwg/react-18/discussions/110).
```

## Adding Strict Effects to Strict Mode

In the future, we'd like to add features to React which would allow a component to mount without immediately creating effects (such as pre-rendering), or to destroy effects in already-mounted components (such as when a component becomes no longer visible). These features will add better performance and resource management out-of-the-box to React, but require effects to be decoupled from the component lifecycle and resilient to being created and destroyed multiple times in a component.

To help surface these issues, React 18 introduced Strict Effects to Strict Mode. With Strict Effects, React will automatically destroy and re-create every effect in development, whenever a component mounts.

For more information, see the Working Group posts for [Adding Strict Effects to Strict Mode](https://github.com/reactwg/react-18/discussions/19) and [How to Support Strict Effects](https://github.com/reactwg/react-18/discussions/18).

## Testing

TODO:
* `IS_REACT_ACT_ENVIRONMENT` https://github.com/reactwg/react-18/discussions/102

## Dropping support for Internet Explorer

> In this release, React dropping support for Internet Explorer, which is going end of life on June 15, 2022. We’re making this change now because new features introduced in React 18 are built using modern browser features such as microtasks which cannot be adequately polyfilled in IE.

Reviewer: ```suggestion
In this release, React is dropping support for Internet Explorer, which is going end of life on June 15, 2022. We’re making this change now because new features introduced in React 18 are built using modern browser features such as microtasks which cannot be adequately polyfilled in IE.
```

> In this release, React dropping support for Internet Explorer, which is going end of life on June 15, 2022. We’re making this change now because new features introduced in React 18 are built using modern browser features such as microtasks which cannot be adequately polyfilled in IE.

Reviewer: I would drop a little note here linking to https://docs.microsoft.com/en-us/lifecycle/faq/internet-explorer-microsoft-edge#what-is-the-lifecycle-policy-for-internet-explorer- because "end-of-life" is not entirely accurate.

* `useSyncExternalStore` is a new hook that allows external stores to support concurrent reads by forcing updates to the store to be synchronous. This new API is recommended for any library that integrates with state external to React. For more information, see the [useSyncExternalStore overview post](https://github.com/reactwg/react-18/discussions/70) and [useSyncExternalStore API details](https://github.com/reactwg/react-18/discussions/86).
* `useInsertionEffect` is a new hook that allows CSS-in-JS libraries to address performance issues of injecting styles in render. This hook will run after the DOM is mutated, but before layout effects read the new layout. This solves an issue that already exists in React 17 and below, but is even more important in React 18 because React yields to the browser during concurrent rendering, giving it a chance to recalculate layout. For more information, see the [Library Upgrade Guide for `<style>`](https://github.com/reactwg/react-18/discussions/110).

React 18 also introduces new APIs for concurrent rendering such as `startTransition` and `useDeferredValue`, which we will share more about in the upcoming stable release post.

> ## Adding Strict Effects to Strict Mode

Reviewer 2: I noticed the capitalization of these headers is inconsistent. Most our previous blog posts use title case, I think, but not always. As long as it's consistent within the post I think that's fine.

Author: Switched everything to title case 👍

});

// After
function AppWithCallbackAfterRender() {
  useEffect(() => {
    console.log('rendered');
  });

  return <App tab="home" />
}

const container = document.getElementById('app');
const root = ReactDOM.createRoot(container);
root.render(<AppWithCallbackAfterRender />);
```

Finally, to upgrade `ReactDOM.hydrate`:

```js
// Before
const container = document.getElementById('app');
ReactDOM.hydrate(<App tab="home" />, container);

// After
const container = document.getElementById('app');
const root = ReactDOM.hydrateRoot(container, <App tab="home" />);
// Unlike with createRoot, you don't need a separate root.render() call here.
```

For more information, see the [working group discussion here](https://github.com/reactwg/react-18/discussions/5).

## Updates to Server Rendering APIs

In this release, we’re revamping our server rendering APIs to support Suspense on the server and Streaming SSR.

As part of these changes, the following APIs will continue working, but with limited support for Suspense. Instead of erroring, these APIs will server render the nearest fallback for any subtree that suspends, and retry rendering on the client:

* `renderToString`

> * `renderToStaticMarkup`

Reviewer 3: renderToStaticMarkup won't "retry rendering on the client" because there's no client in this move, so I guess it just leaves it in permanent fallback.

## Installing

To install the latest React 18 RC, use the `@rc` tag:

```bash
npm install react@rc react-dom@rc
```

Or if you’re using yarn:

```bash
yarn add react@rc react-dom@rc
```

## New Client Root APIs

When you first install React 18, you will see a warning in the console:

![Screenshot of React error with text "Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17."](../images/blog/react-v18-rc/reactdom_warning.png)

React 18 introduces a new root API which provides better ergonomics for managing roots. The new root API also enables the new concurrent renderer, which allows you to opt-into concurrent features.

```js
// Before
import { render } from 'react-dom';
const container = document.getElementById('app');
render(<App tab="home" />, container);

// After
import { createRoot } from 'react-dom/client';
const container = document.getElementById('app');
const root = createRoot(container);
root.render(<App tab="home" />);

```

We’ve also changed `unmountComponentAtNode` to `root.unmount`:

```js
// Before

> unmountComponentAtNode(containter);

Reviewer 4: ```suggestion
unmountComponentAtNode(container);
```

Author: Nice catch, thanks!

// After
root.unmount();
```

And removed the callback from `render`:

```js
// Before
import { render } from 'react-dom';
const container = document.getElementById('app');
render(<App tab="home" />, container, () => {
  console.log('rendered');
});

// After
function AppWithCallbackAfterRender() {
  useEffect(() => {
    console.log('rendered');
  });

  return <App tab="home" />
}

import { createRoot } from 'react-dom/client';
const container = document.getElementById('app');
const root = createRoot(container);

> root.render(<AppWithCallbackAfterRender />);

Reviewer 2: What do you think about deleting this section? It's such a rarely used feature. We can put it in the docs somewhere but I don't think it needs to be in the main upgrade guide.
